### PR TITLE
Allow partial sync with multiple paths

### LINKED ISSUE #164: Allow multiple paths in partial stream sync
Right now the `sync` parameter takes a single path
There is no reason why we couldn't take multiple paths and it opens up scope to optimise build performance by only syncing relevant directories

Considerations:
* Backwards compatability (maybe not necessary with a clear message on the changelog)
* Sync path is used to filter unshelve of pending cl

### PR DESCRIPTION
**Pull request by ca-johnson**
Resolves #164 

Permit specifying a list of paths to sync, instead of limiting it to just one.

* Add utility function `list_from_env_array` which can read either a single value (backwards compatibility) or an array of values from env vars like `ENV_0`, `ENV_1` etc
* Read sync paths via this utility
* Update sync/unshelve to consider each path
* Add test coverage

# Testing strategy
Unit, integration, test this branch in production

### CODE REVIEW

File: python/perforce.py [hunk 1]
```diff
@@ -21,11 +21,15 @@ def __init__(self, root=None, view=None, stream=None,
         root: Directory in which to create the client workspace
         view: Client workspace mapping
         stream: Client workspace stream. Overrides view parameter.
+        sync: List of paths to sync. Defaults to entire view.
+        client_opts: Additional options to add to client. (e.g. allwrite)
+        parallel: How many threads to use for parallel sync.
```
Comment (ca-johnson) [inline python/perforce.py:26]: Driveby, document undocumented params

File: python/perforce.py [hunk 2]
```diff
@@ -21,11 +21,15 @@ def __init__(self, root=None, view=None, stream=None,
...
         """
         self.root = os.path.abspath(root or '')
         self.stream = stream
         self.view = self._localize_view(view or [])
-        self.sync_paths = sync or '//...'
+        self.sync_paths = sync or ['//...']
+        assert isinstance(self.sync_paths, list)
```
Comment (ca-johnson) [inline python/perforce.py:32]: Will never fail other than when authoring unit tests which pass the wrong type - avoids confusion when making this mistake.

Ideally we would add some types to this class instead

File: python/perforce.py [hunk 3]
```diff
@@ -176,9 +180,10 @@ def sync(self, revision=None):
         """Sync the workspace"""
         self._setup_client()
         self.revert()
+        sync_files = ['%s%s' % (path, revision or '') for path in self.sync_paths]
         result = self.perforce.run_sync(
             '--parallel=threads=%s' % self.parallel,
-            '%s%s' % (self.sync_paths, revision or ''),
+            *sync_files,
```
Comment (ca-johnson) [inline python/perforce.py:186]: Expands list into params, e.g.:

```
l = [1, 2, 3]
print(l) => [1, 2, 3]
print(*l) => 1 2 3 # i.e. print(1, 2, 3)
```

File: python/buildkite.py [hunk 4]
```diff
@@ -27,12 +27,28 @@ def get_env():
             env[p4var] = plugin_value
     return env
 
+def list_from_env_array(var):
+    """Read list of values from either VAR or VAR_1, VAR_2 etc"""
```
Comment (ca-johnson) [inline python/buildkite.py]:
```suggestion
    """Read list of values from either VAR or VAR_0, VAR_1 etc"""
```

File: plugin.yml [hunk 5]
```diff
@@ -16,7 +16,7 @@ configuration:
     stream:
       type: string
     sync:
-      type: string
+      type: array
```
Comment (ca-johnson) [inline plugin.yml:19]: This doesnt have any actual functional effect afaik - I think buildkite have some plans to auto-generate docs from this file

**Review by improbable-mattchurch — APPROVED** [review]
